rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala

```diff
@@ -0,0 +1,173 @@
+package com.stripe.rainier.core
+
+import com.stripe.rainier.sampler._
+import com.stripe.rainier.compute._
+
+//implements Simulation-Based Calibration from https://arxiv.org/abs/1804.06788
+final case class SBC[T](priorGenerators: Seq[Generator[Double]],
+                        priorParams: Seq[Real],
+                        posterior: RandomVariable[(Distribution[T], Real)]) {
+
+  import SBC._
+
+  require(priorParams.forall(_.variables.size == 1))
+  val priorGenerator = Generator.traverse(priorGenerators)
+  val emptyEvaluator = new Evaluator(Map.empty)
+
+  def prepare(sampler: Sampler, warmupIterations: Int, syntheticSamples: Int)(
+      implicit rng: RNG): Run = {
+    val initialThin = 2
+    val (_, maxRHat, minEffectiveSampleSize) =
+      sample(sampler, warmupIterations, syntheticSamples, initialThin)
+    val thin =
+      if (minEffectiveSampleSize < Samples)
+        Math
+          .ceil(initialThin.toDouble * Samples / minEffectiveSampleSize)
+          .toInt
+      else
+        initialThin
+    Run(sampler, warmupIterations, syntheticSamples, thin, maxRHat)
+  }
+
+  private def sample(sampler: Sampler,
+                     warmupIterations: Int,
+                     syntheticSamples: Int,
+                     thin: Int)(implicit rng: RNG): (Int, Double, Double) = {
+    val trueValues = priorGenerator.get(rng, emptyEvaluator)
+    implicit val trueEval = new Evaluator(priorParams.zip(trueValues).toMap)
+    val trueOutput = posterior.map(_._2).get
+
+    val syntheticValues =
+      posterior.map(_._1.generator.repeat(syntheticSamples)).get
+    val model = posterior.flatMap {
+      case (d, r) =>
+        d.fit(syntheticValues).map { _ =>
```

Comment:
The computation graph (I don't know what to call it) style gets a little mind-bending for me here, since at first it looks like we have the value of `r` before we actually do the `fit` that constrains it. But I figured out what's happening: the `Real` `r` doesn't have a numerical value within this case statement -- we're just building up DAG that links the value of `r` with the fit of `d` to `syntheticValues`.

Reply:
Yep, exactly.